### PR TITLE
feat(broker): a message can be published idempotent

### LINKED ISSUE #1012: A message can be published idempotent
AT:
* a message can be published with an id to ensure that the message is published only once (idempotent)
* when a message is published with the same id, name and correlation-key then it is rejected

Context:
* is some case it is important that a message is published/correlated only once (e.g. for a message start event)
* in order to make the publication easier (i.e. no need to store which message is successfully published), Zeebe take care of this
* a generic publisher for Kafka can use the Kafka's message position as id

### PR DESCRIPTION
**Pull request by saig0**
* new field 'messageId' in message record
* reject message if another message with the same id, name and
correlation-key is already published

closes #1012

### CODE REVIEW

**Review by npepinpe — COMMENTED** [review]
- Do we expect most messages to have an ID?
- What is the use case of a message without an ID?
- What is the use of the ID? The implementation of `hasMessage` checks if id, name, and correlation key are the same...would name and correlation key alone be enough to identify an ID?
- We should also test that if the ID is null, then it doesn't matter if you publish a message with same name/correlation key (unless that's not intended behaviour)

File: broker-core/src/main/java/io/zeebe/broker/subscription/message/PublishMessageProcessor.java [hunk 1]
```diff
@@ -27,40 +26,68 @@
...
   public void processRecord(TypedRecord<MessageRecord> record) {
-    // a message is always accepted
+    isDuplicate = false;
+
+    final MessageRecord message = record.getValue();
+    entry =
+        new MessageEntry(
+            bufferAsString(message.getName()),
+            bufferAsString(message.getCorrelationKey()),
+            message.getPayload().byteArray(),
```
[the remark this answers is not in the document]
Comment (saig0) [reply broker-core/src/main/java/io/zeebe/broker/subscription/message/PublishMessageProcessor.java]: :roll_eyes:  809819a

**Comment by saig0** [comment]
@npepinpe, I would not assume that most messages have an id. It's an optional feature that you can use if you want to ensure that the message is only published once. It is interesting when you publish messages from a messaging system like Kafka into Zeebe. In other cases, I think you don't need/want that.

For example, an application sends an "order updated" event every time when the order is changed. The event name and the correlation key are always the same. In order not to force the application to provide a unique id, the application can publish the messages without id.

Currently, the `messageId` can't be `null`. It must be a string. Otherwise, the message is not accepted because the de-serialization fails.

Does this make sense for you?

File: broker-core/src/main/java/io/zeebe/broker/subscription/message/state/MessageDataStore.java [hunk 2]
```diff
@@ -32,6 +32,18 @@ public void addMessage(MessageEntry message) {
     getData().messages.add(message);
   }
 
+  public boolean hasMessage(MessageEntry message) {
+    return getData()
+        .messages
+        .stream()
+        .anyMatch(
+            m ->
+                m.getId() != null
```
Comment (npepinpe) [inline broker-core/src/main/java/io/zeebe/broker/subscription/message/state/MessageDataStore.java]: If a message can't be null, shouldn't we check for empty string here then? My understanding of `BaseProperty.hasValue` in this case means all messages will always have a value, and therefore every `MessageEntry` will be created with either a normal ID, or an empty string, meaning two messages with empty string IDs, same name, same correlation key, would be the same?

**Review by npepinpe — COMMENTED** [review]
Yes, it's clear, but in the code we do check for null (indirectly I know).

File: broker-core/src/main/java/io/zeebe/broker/subscription/message/state/MessageDataStore.java (hunk 2 above)
Comment (saig0) [reply broker-core/src/main/java/io/zeebe/broker/subscription/message/state/MessageDataStore.java]: You're right. I fixed the check if the message id is set -> a099b72

**Review by npepinpe — APPROVED** [review]

**Comment by menski** [comment]
bors retry

**Pull request merged**